### PR TITLE
Allow AWS to handle encryption

### PR DESCRIPTION
**Pull request by dora-korpar**
On a PUT request to AWS backend, AWS will handle encryption.

### CODE REVIEW

**Comment by dora-korpar** [comment]
@ironman-machine go

File: lib/data/external/AwsClient.js [hunk 1]
```diff
@@ -42,7 +45,6 @@ class AwsClient {
                }
                // because of encryption the ETag here could be
```
Comment (LaurenSpiegel) [inline lib/data/external/AwsClient.js]: I don't think this comment is relevant anymore and we shouldn't save the etag.  Let's be consistent with what we do for the other backends.

File: lib/data/external/AwsClient.js [hunk 2]
```diff
@@ -24,15 +24,18 @@ class AwsClient {
...
             Key: awsKey,
             Body: stream,
             Metadata: keyContext.metaHeaders,
             ContentLength: size,
-        },
+        };
+        if (keyContext.cipherBundle) {
+            uploadParams.ServerSideEncryption = 'aws:kms';
+        }
+
```
Comment (LaurenSpiegel) [inline lib/data/external/AwsClient.js]: please add a test where you put an object to the aws backend in an encrypted bucket and then you get the object directly from aws to confirm that you can retrieve it and that it sends back the proper encryption headers.

**Comment by dora-korpar** [comment]
@ironman-machine go

File: lib/data/locationConstraintParser.js [hunk 3]
```diff
@@ -53,6 +53,8 @@ function parseLC() {
                         keepAlive: true,
                     }),
                 },
+                // needed for encryption
+                signatureVersion: 'v4',
```
Comment (LaurenSpiegel) [inline lib/data/locationConstraintParser.js]: does this mess with being able to stream objects?  see not above

**Comment by dora-korpar** [comment]
@ironman-machine go

File: tests/functional/aws-node-sdk/lib/utility/createEncryptedBucket.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+const getConfig = require('../../test/support/config');
+const childProcess = require('child_process');
```
Comment (electrachong) [inline tests/functional/aws-node-sdk/lib/utility/createEncryptedBucket.js]: can put the external modules on top

```
const childProcess = require('child_process');
const Promise = require('bluebird');
const getConfig = require('../../test/support/config');
```

File: tests/functional/aws-node-sdk/lib/utility/createEncryptedBucket.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        args = args.concat('-s');
+    }
+    const body = [];
+    const child = childProcess.spawn(args[0], args)
+    .on('exit', () => {
+        const hasSucceed = body.join('').split('\n').find(item => {
+            const json = safeJSONParse(item);
+            const test = !(json instanceof Error) && json.name === 'S3' &&
+                json.statusCode === 200;
+            if (test) {
```
Comment (bennettbuchanan) [inline tests/functional/aws-node-sdk/lib/utility/createEncryptedBucket.js]: Minor: can just return `test` here.

File: tests/functional/aws-node-sdk/test/multipleBackend/put.js [hunk 6]
```diff
@@ -44,7 +46,13 @@ function awsGetCheck(objectKey, s3MD5, awsMD5, cb) {
         (err, res) => {
             assert.strictEqual(err, null, 'Expected success, got error ' +
             `on direct AWS call: ${err}`);
-            assert.strictEqual(res.ETag, `"${awsMD5}"`);
+            if (process.env.ENABLE_KMS_ENCRYPTION === 'true') {
+                // doesn't check ETag because it's different
+                // with every PUT with encryption
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/multipleBackend/put.js]: is the etag the md5 of the encrypted data?

File: tests/functional/aws-node-sdk/test/multipleBackend/put.js [hunk 7]
```diff
@@ -4,6 +4,8 @@ const withV4 = require('../support/withV4');
 const BucketUtility = require('../../lib/utility/bucket-util');
 const { config } = require('../../../../../lib/Config');
 const { getRealAwsConfig } = require('../support/awsConfig');
+const { createEncryptedBucketPromise } =
+    require('../../lib/utility/createEncryptedBucket');
 
 const awsBucket = 'multitester555';
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/multipleBackend/put.js]: can we pull this from the config so it's not hard coded in many places?

File: tests/functional/aws-node-sdk/test/multipleBackend/put.js (hunk 6 above)
Comment (dora-korpar) [inline tests/functional/aws-node-sdk/test/multipleBackend/put.js]: AWS says it's not the md5 of the object data, but doesn't specify what it actually is

Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/multipleBackend/put.js]: ok

**Comment by dora-korpar** [comment]
@ironman-machine go

**Comment by dora-korpar** [comment]
@ironman-machine go2

**Comment by dora-korpar** [comment]
@ironman-machine go3

**Pull request merged**